File: cargoLedger/hasher.py
import hashlib
import os
import fnmatch
from lxml import etree

DIRECTORY = "./demo_order"
BUNDLE_SIZE = 50

# Helper function to hash a text using SHA-3


def hash_sha3(text):
    return hashlib.sha3_256(text.encode('utf-8')).hexdigest()

# Recursive function to traverse the XML tree and hash each field


def traverse_and_hash(node, parent_tag=None):
    hash_dict = {}
    for child in node:
        if len(child) > 0:
            child_hash_dict = traverse_and_hash(child, parent_tag=child.tag)
            hash_dict.update(child_hash_dict)
        else:
            hashed_value = hash_sha3(child.text)
            key = f"{parent_tag}-{child.tag}" if parent_tag else child.tag
            hash_dict[key] = hashed_value

    return hash_dict


# Function to process an XML file


def process_xml_file(file_path):
    with open(file_path, 'rb') as f:  # Read the file as bytes
        xml_data = f.read()

    root = etree.fromstring(xml_data)
    xml_hash_dict = traverse_and_hash(root)
    # Concatenate the hashes in the xml_hash_dict
    concatenated_hashes = ''.join(xml_hash_dict.values())

    # Compute the overarching hash
    overarching_hash = hash_sha3(concatenated_hashes)

    return overarching_hash

# Function to process XML files in a directory and its subdirectories


def process_xml_files_in_directory(directory):
    hashes = {}
    for root, dirs, files in os.walk(directory):
        for file in fnmatch.filter(files, '*.xml'):
            file_path = os.path.join(root, file)

            process_xml_hash = process_xml_file(file_path)
            hashes[file_path] = process_xml_hash
    print(hashes)
    return hashes


# Start processing XML files in the directory and its subdirectories


def main():
    hash_per_file = process_xml_files_in_directory(DIRECTORY)
    hashes = list(hash_per_file.values())
    length = BUNDLE_SIZE
    concat_hash = ""

    if len(hashes) < BUNDLE_SIZE:
        length = len(hashes)

    for i in range(length):
        concat_hash += hashes[i]

    overreaching_hash = hash_sha3(concat_hash)
    print(overreaching_hash)
    return overreaching_hash


if __name__ == "__main__":
    main()

File: cargoLedger/hasher.js
const crypto = require("crypto-js");
const xml2js = require("xml2js");
const fs = require("fs");
const path = require("path");

const DIRECTORY = "demo_order";
const BUNDLE_SIZE = 50;

// Helper function to hash a text using SHA-3
function hashSha3(text) {
  return crypto.SHA3(text, { outputLength: 256 }).toString(crypto.enc.Hex);
}

// Recursive function to traverse the XML tree and hash each field
function traverseAndHash(node, parentTag = null) {
  let hashDict = {};
  for (let childTag in node) {
    if (typeof node[childTag] === "object") {
      let childHashDict = traverseAndHash(
        node[childTag],
        (parentTag = childTag)
      );
      hashDict = { ...hashDict, ...childHashDict };
    } else {
      let hashedValue = hashSha3(node[childTag]);
      let key = parentTag ? `${parentTag}-${childTag}` : childTag;
      hashDict[key] = hashedValue;
    }
  }
  return hashDict;
}

// Function to process an XML file
async function processXmlFile(filePath) {
  let xmlData = fs.readFileSync(filePath);
  let parsedData = await xml2js.parseStringPromise(xmlData);

  let xmlHashDict = traverseAndHash(parsedData);
  // Concatenate the hashes in the xmlHashDict
  let concatenatedHashes = Object.values(xmlHashDict).join("");

  // Compute the overarching hash
  let overarchingHash = hashSha3(concatenatedHashes);

  return overarchingHash;
}

// Function to process XML files in a directory and its subdirectories
async function processXmlFilesInDirectory(directory) {
  let hashes = {};

  try {
    const items = await fs.promises.readdir(directory, { withFileTypes: true });

    // Listing all items using forEach
    for (const item of items) {
      // Check if the item is a directory
      if (item.isDirectory()) {
        // Process the directory recursively
        let subDirHashes = await processXmlFilesInDirectory(
          path.join(directory, item.name)
        );
        hashes = { ...hashes, ...subDirHashes };
      }

      // Check if the file has a .xml extension
      if (item.isFile() && item.name.endsWith(".xml")) {
        // Process the file
        let filePath = path.join(directory, item.name);
        // await till the file is processed

        let hash = await processXmlFile(filePath);
        hashes[filePath] = hash;

        // Move the file to the "processed" directory
        // moveXmlFilesToProcessedDirectory(filePath);
      }
    }
  } catch (error) {
    console.log("Unable to scan directory: " + error);
  }
  return hashes;
}

// Function to move XML files to processed directory
async function moveXmlFilesToProcessedDirectory(filePath) {
  try {
    // processed directory
    const baseProcessedDir = path.join(path.dirname(DIRECTORY), "processed");

    // Get the relative directory path from the current working directory
    const fileDir = path.dirname(filePath);
    const relativeDirPath = path.relative(DIRECTORY, fileDir);

    // Generate a unique filename using the current timestamp
    const originalFileName = path.basename(filePath, ".xml");
    const timestamp = new Date().getTime();
    const newFileName = `${originalFileName}.processed(${timestamp}).xml`;

    // Create the new file path with the relative directory hierarchy
    const processedDir = path.join(baseProcessedDir, relativeDirPath);
    const newFilePath = path.join(processedDir, newFileName);

    // Create the "processed" directory if it doesn't exist
    if (!fs.existsSync(processedDir)) {
      fs.mkdirSync(processedDir);
    }

    // Move the file to the "processed" directory
    await fs.promises.rename(filePath, newFilePath);
  } catch (error) {
    console.log("Unable to move file to processed directory: " + error);
  }
}

// Start processing XML files in the directory and its subdirectories
async function hasher() {
  let directoryPath = path.join(__dirname, DIRECTORY);
  let hashPerFile = await processXmlFilesInDirectory(directoryPath);
  let hashes = Object.values(hashPerFile);
  let length = BUNDLE_SIZE;
  let concatHash = "";

  if (hashes.length < BUNDLE_SIZE) {
    length = hashes.length;
  }

  for (let i = 0; i < length; i++) {
    concatHash += hashes[i];
  }
  let overreachingHash = hashSha3(concatHash);
  console.log(overreachingHash);
  return overreachingHash;
}

module.exports = hasher;
